### DISCUSSION TITLE
An explanation for the source code of finding the alignment path in GlowTTS?

### DISCUSSION THREAD

**Discussion by alifarrokh** [coqui-ai/TTS#1952] [General Q&A] [answered]
Hi. I'm reading the source code of `GlowTTS` model for educational purposes. One of the sections that I can't really understand is where we try to find the alignment path using Monotonic Alignment Search in the training phase. Could anyone please explain me the following lines of [code](https://github.com/coqui-ai/TTS/blob/371772c355124556531c089770d6a8b110daf160/TTS/tts/models/glow_tts.py#L241)?

```python
'''
TTS/TTS/tts/models/glow_tts.py
lines 241 to 251
'''
# find the alignment path
with torch.no_grad():
    o_scale = torch.exp(-2 * o_log_scale)
    logp1 = torch.sum(-0.5 * math.log(2 * math.pi) - o_log_scale, [1]).unsqueeze(-1)  # [b, t, 1]
    logp2 = torch.matmul(o_scale.transpose(1, 2), -0.5 * (z**2))  # [b, t, d] x [b, d, t'] = [b, t, t']
    logp3 = torch.matmul((o_mean * o_scale).transpose(1, 2), z)  # [b, t, d] x [b, d, t'] = [b, t, t']
    logp4 = torch.sum(-0.5 * (o_mean**2) * o_scale, [1]).unsqueeze(-1)  # [b, t, 1]
    logp = logp1 + logp2 + logp3 + logp4  # [b, t, t']
    attn = maximum_path(logp, attn_mask.squeeze(1)).unsqueeze(1).detach()
y_mean, y_log_scale, o_attn_dur = self.compute_outputs(attn, o_mean, o_log_scale, x_mask)
attn = attn.squeeze(1).permute(0, 2, 1)
```

Thanks in advance.

**Comment by erogol** [comment]
This code computes the affinity between the text encoder outputs and the decoder outputs in general by just computing log scale Gaussian factors.

**Reply by alifarrokh** [reply]
Can you please provide a formula or a reference where I can read about it?

**Reply by lexkoro** [reply]
@alifarrokh GlowTTS paper https://arxiv.org/abs/2005.11129

**Reply by alifarrokh** [reply]
I've already read this paper, yet I can't fully understand this section.

**Comment by p0p4k** [comment, marked as answer]
Hello, I can help you with this. Add me on Gitter @p0p4k
While the code you pasted is actually the log N (z, mu, sigma) calculation first, then it sends to  [maximum path function](https://github.com/coqui-ai/TTS/blob/371772c355124556531c089770d6a8b110daf160/TTS/tts/utils/helpers.py#L174-L238). which implements the following part of the paper.
![image](https://user-images.githubusercontent.com/8834712/190064107-a3b79c83-1df6-4e3a-8740-7e7bc28f586c.png)
(refer to #1799 for related discussions.)
